dbms/src/Interpreters/InterpreterInsertQuery.cpp

```diff
@@ -99,73 +99,99 @@ BlockIO InterpreterInsertQuery::execute()
+        ///       Otherwise we'll get duplicates when MV reads same rows again from Kafka.
+        if (table->noPushingToViews() && !no_destination)
+            out = table->write(query_ptr, context);
+        else
+            out = std::make_shared<PushingToViewsBlockOutputStream>(query.database, query.table, table, context, query_ptr, no_destination);
 
-    auto out_wrapper = std::make_shared<CountingBlockOutputStream>(out);
-    out_wrapper->setProcessListElement(context.getProcessListElement());
-    out = std::move(out_wrapper);
+        /// Do not squash blocks if it is a sync INSERT into Distributed, since it lead to double bufferization on client and server side.
+        /// Client-side bufferization might cause excessive timeouts (especially in case of big blocks).
+        if (!(context.getSettingsRef().insert_distributed_sync && table->isRemote()) && !no_squash)
+        {
+            out = std::make_shared<SquashingBlockOutputStream>(
+                out, out->getHeader(), context.getSettingsRef().min_insert_block_size_rows, context.getSettingsRef().min_insert_block_size_bytes);
+        }
 
-    BlockIO res;
+        /// Actually we don't know structure of input blocks from query/table,
+        /// because some clients break insertion protocol (columns != header)
+        out = std::make_shared<AddingDefaultBlockOutputStream>(
+            out, query_sample_block, out->getHeader(), table->getColumns().getDefaults(), context);
+
+        if (const auto & constraints = table->getConstraints(); !constraints.empty())
+            out = std::make_shared<CheckConstraintsBlockOutputStream>(query.table,
+             out, query_sample_block, table->getConstraints(), context);
+
+        auto out_wrapper = std::make_shared<CountingBlockOutputStream>(out);
+        out_wrapper->setProcessListElement(context.getProcessListElement());
+        out = std::move(out_wrapper);
+        out_streams.emplace_back(std::move(out));
+    }
 
     /// What type of query: INSERT or INSERT SELECT?
     if (query.select)
     {
-        /// Passing 1 as subquery_depth will disable limiting size of intermediate result.
-        InterpreterSelectWithUnionQuery interpreter_select{query.select, context, SelectQueryOptions(QueryProcessingStage::Complete, 1)};
+        for (auto & in_stream : in_streams)
+        {
+            in_stream = std::make_shared<ConvertingBlockInputStream>(
+                context, in_stream, out_streams.at(0)->getHeader(), ConvertingBlockInputStream::MatchColumnsMode::Position);
+        }
 
-        /// BlockIO may hold StoragePtrs to temporary tables
-        res = interpreter_select.execute();
-        res.out = nullptr;
+        Block in_header = in_streams.at(0)->getHeader();
+        if (in_streams.size() > 1)
+        {
+            for (size_t i = 1; i < in_streams.size(); ++i)
+                assertBlocksHaveEqualStructure(in_streams[i]->getHeader(), in_header, "INSERT SELECT");
```

Comment:
Actually, `executeWithMultipleStreams` already checks that streams have equal headers. And add conversion if not.
But let's leave this check here just in case.